src/submission/exceptions/exceptions.ts

```diff
@@ -260,3 +274,69 @@ export const checkForProgramAndEntityExceptions = async ({
 	});
 	return { filteredErrors, normalizedRecord };
 };
+
+/**
+ * Collect all exception records related to a set of Donors
+ *
+ * @param programId
+ * @param filters
+ * @returns [ ExceptionRecords ]
+ */
+export async function getExceptionManifestRecords(
+	programId: string,
+	filters: { donorIds: number[]; submitterDonorIds: string[] },
+): Promise<ExceptionManifestRecord[]> {
+	const { donorIds, submitterDonorIds: querySubmitterIds } = filters;
+
+	const donorsByDonorId = await getDonorsByIds(donorIds);
+	const donorsBySubmitterId = (await findDonorsBySubmitterIds(programId, querySubmitterIds)) || [];
+
+	const donors = [...donorsByDonorId, ...donorsBySubmitterId].filter(notEmpty).filter(
+		(donorRecord, index, donorArray) =>
+			// Filter duplicates
+			index === donorArray.findIndex((donor) => donor.donorId === donorRecord.donorId),
+	);
+
+	// Exceptions only store submitterIds, so all submitterIds have to be collected before we can filter exceptions
+	const submitterDonorIds = donors.map((donor) => donor.submitterId);
+
+	const { programException, entityException: entityPropertyException } = await queryForExceptions(
+		programId,
+	);
+
+	const missingEntityException = await getByProgramId(programId);
+
+	const programExceptions = programException?.exceptions || [];
+
+	const programExceptionDisplayRecords: ProgramPropertyExceptionRecord[] = programExceptions.map(
+		createProgramExceptions(programId),
+	);
+
+	const entityPropertyExceptions: EntityPropertyExceptionRecord[] = entityPropertyException
+		? [
+				...entityPropertyException.specimen,
+				...entityPropertyException.treatment,
+				...entityPropertyException.follow_up,
+		  ]
+				.filter((exceptionRecord) => submitterDonorIds.includes(exceptionRecord.submitter_donor_id))
+				.map(mapEntityExceptionRecords(programId, donors))
+		: [];
+
+	const missingEntityExceptions: MissingEntityExceptionRecord[] = missingEntityException.success
+		? missingEntityException.data.donorSubmitterIds
+				.filter((submitterDonorId) => submitterDonorIds.includes(submitterDonorId))
+				.map((submitterDonorId) => {
+					const exceptionType = MissingEntityExceptionType;
+					const { donorId } = donors.find((donor) => donor.submitterId === submitterDonorId) || {};
+					return { programId, exceptionType, submitterDonorId, donorId };
+				})
+		: [];
+
+	const donorExceptionRecords: ExceptionManifestRecord[] = [
+		...programExceptionDisplayRecords,
+		...entityPropertyExceptions,
+		...missingEntityExceptions,
+	];
+
```

Comment:
not a homogeneous array meaning it's holding different types of objects which I imagine being an issue here: https://github.com/icgc-argo/argo-clinical/pull/1161/files#diff-654b5ddd8e9c70e984a811687ea0784f6467af540e4cc7eb7ff9a831551e7cf7R75 when you try and make a TSV out of it.   (I changed the base branch to this branch so the diff looks right)

I think you'll end up with the union of all columns, which some records won't have any data for.
probably need to return an array of all the same type or an object with the difference types might be work too
```ts
const donorExceptionRecords = {programExceptions: [], entityExceptiosn: []}
```

Reply:
Yes this is correct, however we are only making 1 TSV for all exceptions, so I think we're going to see all columns regardless. I will confirm what's the expected TSV output!